system/masking/include/datamaskingengine.hpp

```diff
@@ -0,0 +1,611 @@
+                {
+                    trace->uerrlog("plugin loader name lookup failed - no entry point name and no default value");
+                    canFind = false;
+                }
+                else
+                    entryPointName = entryPointDefault.c_str();
+            }
+            return (canFind ? find(libraryName, entryPointName) : nullptr);
+        }
+
+    private:
+        using EntryPointMap = std::map<std::string, entry_point_t>;
+        struct Plugin
+        {
+            HINSTANCE library = nullptr;
+            EntryPointMap entryPoints;
+        };
+        using PluginMap = std::map<std::string, Plugin>;
+        Linked<ITraceProducer> trace;
+        PluginMap              plugins;
+        CriticalSection        lock;
+        std::string            libraryXPath;
+        std::string            libraryDefault;
+        std::string            entryPointXPath;
+        std::string            entryPointDefault;
+\
+        entry_point_t find(const char* libraryName, const char* entryPointName)
+        {
+            std::string realName;
+            realName += SharedObjectPrefix;
+            realName += libraryName;
+            realName += SharedObjectExtension;
+            CriticalBlock block(lock);
+            Plugin& plugin = plugins[libraryName];
+            if (!plugin.library)
+            {
+                plugin.library = LoadSharedObject(realName.c_str(), true, false);
+                if (!plugin.library)
+                {
+                    trace->uerrlog("plugin loader lookup failed - cannot load library '%s'", realName.c_str());
+                    return nullptr;
+                }
+            }
+            entry_point_t& entryPoint = plugin.entryPoints[entryPointName];
+            if (!entryPoint)
+            {
+                entryPoint = (entry_point_t)GetSharedProcedure(plugin.library, entryPointName);
+                if (!entryPoint)
+                {
+                    trace->uerrlog("plugin loader lookup failed - library '%s' does not export '%s'", realName.c_str(), entryPointName);
+                    return nullptr;
+                }
+            }
+            return entryPoint;
+        }
+    };
+
+    class CEngine : public CInterfaceOf<IDataMaskingEngine>
+    {
+    protected:
+        class Loader
+        {
+            Linked<ITraceProducer>                     trace;
+            TPluginLoader<DataMaskingPluginEntryPoint> loader;
+        public:
+            Loader(ITraceProducer& _trace)
+                : loader(nullptr, nullptr, "@library", "@entry-point", _trace)
+            {
+                trace.set(&_trace);
+            }
+            IDataMaskingProfileIterator* create(const IPTree& configuration)
+            {
+                Owned<IDataMaskingProfileIterator> plugin(loader.create<IDataMaskingProfileIterator>(configuration, [&](DataMaskingPluginEntryPoint entryPoint) { return entryPoint(configuration, *trace); }));
```

Comment:
I would usually name the instance of an iterator as plural.

Reply:
Singular is correct, because a plugin is by definition a profile iterator. Will change to `profiles` to address confusion.